### ISSUE TITLE
Unable to match requests when using Pyramid's route_prefix

### ISSUE THREAD

**Issue by tholo** [striglia/pyramid_swagger#148]
I have been getting "Could not find a matching Swagger operation for GET request api/v1/users" from one of my ... experiments.

After digging in to it a bit, I found that the problem was either that `route_info['route'].path` in tween.py:get_op_for_request() unexpectedly does *not* start with a "/", *or* that the map used in the `Spec` from bravado_core unexpectedly *does* start the requests with a "/".

This is with Python 3.5.0, pyramid 1.5.7, pyramid-swagger 2.2.2, bravado-core 3.1.1

**Comment by striglia** [comment]
Hey Thorsten,

Sounds like this might just be a mismatch between your specific request URI and the swagger spec for your service. Can you post your full api-docs.json spec as well as the method you were using to query the service (a simple curl works)?

**Comment by tholo** [comment]
The Swagger spec is below. With the proper credential, a command

```sh
curl -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJleHAiOjE0NDc4MTk3ODcsImlhdCI6MTQ0NzgwODk4Nywic3ViIjoxLCJuYmYiOjE0NDc4MDg5ODd9.AAQQvvH4zWn6Ow4O6VC4nQA_2SdnHMIQCxwfoP8jQxz19n1Z3tZkh2NvjntIbFMou4Gk-tIQAa4bD0_4lI8TwV-BAU2mFv0-TnkzjTJ2CHdWYTJhdIFlj7ezQLsbH2cQUNuhc_901Edn5soYBZQQ1Vtt9DNbuoFe3SEsxbBbjEJn-aPy" https://protalk.ngrok.io/api/v1/users
```

will fail. Based on the spec that ... should not be the case, I believe.

Thorsten

```json
{
  "swagger": "2.0",
  "info": {
    "title": "Messaging",
    "description": "Backend for secure messaging application",
    "termsOfService": "https://protalk.ngrok.io/terms/",
    "contact": {
      "name": "Support",
      "url": "https://protalk.ngrok.io/support/",
      "email": "support@example.com"
    },
    "license": {
      "name": "Proprietary",
      "url": "https://protalk.ngrok.io/license"
    },
    "version": "1.0.0"
  },
  "host": "protalk.ngrok.io",
  "basePath": "/api/v1",
  "schemes": ["https"],
  "consumes": [
    "application/json"
  ],
  "produces": [
    "application/json"
  ],
  "responses": {
    "NotFound": {
      "description": "Entity not found.",
      "schema": {
        "$ref": "#/definitions/Error"
      }
    },
    "AuthenticationRequired": {
      "description": "Authentication is required to do that.",
      "schema": {
        "$ref": "#/definitions/Error"
      }
    }
  },
  "securityDefinitions": {
    "messaging_auth": {
      "type": "oauth2",
      "tokenUrl": "https://protalk.ngrok.io/auth/login",
      "flow": "password",
      "scopes": {
      }
    }
  },
  "paths": {
    "/users": {
      "get": {
        "summary": "Get a list of users.",
        "description": "Retrieve a list of all users from the server.",
        "security": [
          {
            "messaging_auth": []
          }
        ],
        "responses": {
          "200": {
            "description": "List of users",
            "schema": {
              "type": "array",
              "items": {
                "$ref": "#/definitions/User"
              }
            }
          },
          "401": {
            "$ref": "#/responses/AuthenticationRequired"
          },
          "default": {
            "description": "Unexpected error",
            "schema": {
              "$ref": "#/definitions/Error"
            }
          }
        }
      }
    },
    "/users/{userId}": {
      "get": {
        "summary": "Get details about user.",
        "description": "Fetch all details for a single user identified by the passed in userId.",
        "operationId": "getUserById",
        "parameters": [
          {
            "in": "path",
            "name": "userId",
            "description": "User to fetch",
            "required": true,
            "type": "integer",
            "format": "int64"
          }
        ],
        "responses": {
          "200": {
            "description": "Information about user",
            "schema": {
              "$ref": "#/definitions/User"
            },
            "examples": {
              "application/json": {
                "userId": 1,
                "username": "user",
                "firstname": "John",
                "lastname": "Doe",
                "email": "john.doe@example.com"
              }
            }
          },
          "401": {
            "$ref": "#/responses/AuthenticationRequired"
          },
          "404": {
            "$ref": "#/responses/NotFound"
          },
          "default": {
            "description": "Unexpected error",
            "schema": {
              "$ref": "#/definitions/Error"
            }
          }
        },
        "security": [
          {
            "messaging_auth": []
          }
        ]
      }
    }
  },
  "definitions": {
    "User": {
      "type": "object",
      "properties": {
        "userId": {
          "type": "integer"
        },
        "username": {
          "type": "string"
        },
        "firstname": {
          "type": "string"
        },
        "lastname": {
          "type": "string"
        },
        "email": {
          "type": "string"
        }
      },
      "required": [
        "userId"
      ],
      "example": {
        "userId": 1,
        "username": "user",
        "firstname": "John",
        "lastname": "Doe",
        "email": "john.doe@example.com"
      }
    },
    "Error": {
      "type": "object",
      "properties": {
        "message": {
          "type": "string"
        }
      },
      "required": [
        "message"
      ],
      "example": {
        "message": "User not found."
      }
    }
  }
}
```

**Comment by striglia** [comment]
Interesting. My educated guess is that we're failing to interpret your spec's use of basePath properly. You're doing it right, so that is probably just a library support issue. 

I see support for this behavior [in bravado-core](https://github.com/Yelp/bravado-core/blob/4840a6e374611bb917226157b5948ee263913abc/bravado_core/spec.py#L152), so I'm unsure why this is failing as written.

@analogue: any ideas here?

@tholo: can you try again without using the basePath feature so we can at least narrow this down to that as an issue?

**Comment by tholo** [comment]
FWIW if I modify tween.py:get_op_for_request() to add a leading '/' if none are present, that does make things work. It just seems ... like the wrong way to fix it. I will try to see what happens if I do not use basePath but fully prefix things.

**Comment by tholo** [comment]
So removing `basePath` and specifying `/api/v1/users` instead of just `/users` (with `basePath`) in `paths` gives me the same error...

Specifically, the `paths` section now starts out with:

```json
  "paths": {
    "/api/v1/users": {
      "get": {
```

and there is no `basePath` setting. The **curl** command from before gives the exact same error message.

**Comment by tholo** [comment]
Let's see what `request` and `route` looks like... Note that this is without using `basePath` still, tho that seems to have changed … nothing.

```python
{'_headers': <webob.headers.EnvironHeaders object at 0x109479198>,
 'environ': {'HTTP_ACCEPT': '*/*',
             'HTTP_AUTHORIZATION': 'Bearer '
                                   'eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9.eyJleHAiOjE0NDc4MTk3ODcsImlhdCI6MTQ0NzgwODk4Nywic3ViIjoxLCJuYmYiOjE0NDc4MDg5ODd9.AAQQvvH4zWn6Ow4O6VC4nQA_2SdnHMIQCxwfoP8jQxz19n1Z3tZkh2NvjntIbFMou4Gk-tIQAa4bD0_4lI8TwV-BAU2mFv0-TnkzjTJ2CHdWYTJhdIFlj7ezQLsbH2cQUNuhc_901Edn5soYBZQQ1Vtt9DNbuoFe3SEsxbBbjEJn-aPy',
             'HTTP_HOST': 'protalk.ngrok.io',
             'HTTP_USER_AGENT': 'curl/7.43.0',
             'HTTP_X_FORWARDED_FOR': '50.255.53.125',
             'PATH_INFO': '/api/v1/users',
             'QUERY_STRING': '',
             'REMOTE_ADDR': '127.0.0.1',
             'REQUEST_METHOD': 'GET',
             'SCRIPT_NAME': '',
             'SERVER_NAME': 'localhost',
             'SERVER_PORT': '6543',
             'SERVER_PROTOCOL': 'HTTP/1.1',
             'SERVER_SOFTWARE': 'waitress',
             'wsgi.errors': <_io.TextIOWrapper name='<stderr>' mode='w' encoding='UTF-8'>,
             'wsgi.file_wrapper': <class 'waitress.buffers.ReadOnlyFileBasedBuffer'>,
             'wsgi.input': <_io.BytesIO object at 0x109473db0>,
             'wsgi.multiprocess': False,
             'wsgi.multithread': True,
             'wsgi.run_once': False,
             'wsgi.url_scheme': 'https',
             'wsgi.version': (1, 0)},
 'invoke_subrequest': <bound method Router.invoke_subrequest of <pyramid.router.Router object at 0x108eff5c0>>,
 'registry': <Registry messaging>,
 'tm': <transaction._manager.ThreadTransactionManager object at 0x1065191c8>}
```

and

```python
{'factory': <class 'messaging.views.api.users.UserCollection'>,
 'generate': <function _compile_route.<locals>.generator at 0x109455378>,
 'match': <function _compile_route.<locals>.matcher at 0x109455400>,
 'name': 'rest-UserCollection',
 'path': 'api/v1/users',
 'pattern': 'api/v1/users',
 'predicates': [],
 'pregenerator': None}
```

For good measure I also dumped the Spec:_request_to_op_map:

```python
{('delete', '/api/v1/users/{userId}'): Operation(deleteUser),
 ('get', '/api/v1/users'): Operation(get_users),
 ('get', '/api/v1/users/{userId}'): Operation(getUserById),
 ('post', '/api/v1/users'): Operation(post_users),
 ('put', '/api/v1/users/{userId}'): Operation(updateUser)}
```

**Comment by analogue** [comment]
So your ``route.path`` is not prefixed with a ``/``.  When I run my pyramid app, my ``route.path`` is prefixed with a ``/``.  Wonder why that is?

**Comment by tholo** [comment]
I don't know. But that is why I initially said that I could add that in in `get_op_for_request` (but that that felt like the wrong way to go about it)...

**Comment by tholo** [comment]
Python 2.7 makes no difference. I do wonder, tho... I am also using rest_toolkit (0.7 right now, since newer requires Pyramid>=1.6a), maybe it does ... funky stuff when setting up routes, and that is where this whole issues originates...

**Comment by tholo** [comment]
That does not appear to be it, either... Here's what `proutes` says:

```text
Name                                  Pattern                   View                                                                           Method
----                                  -------                   ----                                                                           ------
pyramid_swagger.swagger20.api_docs    /swagger.json             pyramid_swagger.api.view_for_swagger_dot_json                                  *
rest-UserCollection                   /api/v1/users             pyramid.config.views.<pyramid.config.views.MultiView object at 0x109206a58>    *
rest-UserCollection                   /api/v1/users             rest_toolkit.views.default_options_view                                        OPTIONS
rest-UserCollection                   /api/v1/users             rest_toolkit.views.default_delete_view                                         DELETE
rest-UserCollection                   /api/v1/users             rest_toolkit.views.default_get_view                                            GET
rest-UserCollection                   /api/v1/users             rest_toolkit.views.default_patch_view                                          PATCH
rest-UserCollection                   /api/v1/users             rest_toolkit.views.default_post_view                                           POST
rest-UserCollection                   /api/v1/users             rest_toolkit.views.default_put_view                                            PUT
rest-UserResource                     /api/v1/users/{userId}    pyramid.config.views.<pyramid.config.views.MultiView object at 0x109212208>    *
rest-UserResource                     /api/v1/users/{userId}    rest_toolkit.views.default_options_view                                        OPTIONS
rest-UserResource                     /api/v1/users/{userId}    rest_toolkit.views.default_delete_view                                         DELETE
rest-UserResource                     /api/v1/users/{userId}    rest_toolkit.views.default_get_view                                            GET
rest-UserResource                     /api/v1/users/{userId}    rest_toolkit.views.default_patch_view                                          PATCH
rest-UserResource                     /api/v1/users/{userId}    rest_toolkit.views.default_post_view                                           POST
rest-UserResource                     /api/v1/users/{userId}    rest_toolkit.views.default_put_view                                            PUT
```

**Comment by analogue** [comment]
If the missing leading ``/`` on ``route.path`` is the only problem, it may be OK to accept both forms when matching to an operation because of this [verbage](http://docs.pylonsproject.org/projects/pyramid/en/latest/narr/urldispatch.html#route-pattern-syntax).  FYI, ``route.pattern`` is effectively ``route.path`` according to this [code](https://github.com/Pylons/pyramid/blob/1.5.7/pyramid/urldispatch.py#L34). I'm certainly no pyramid expert - someone else want to chime in on this?  @striglia ?

**Comment by tholo** [comment]
Which is basically what I accomplished by adding it, if not there, in `tween.py:get_op_for_request`. Which makes things work, at any rate, I just … didn't like it.

**Comment by striglia** [comment]
Well this got fascinating/ridiculous in a hurry. Something I clearly haven't seen before is going on.... @tholo perhaps we should just pastebin a minimal pyramid app that reproduces this problem? Something like one endpoint with no arguments, and the associated Swagger spec?

**Comment by tholo** [comment]
The plot thickens (again?)...

I think this may be in part due to me using `config.include(..., prefix='/api/v1')` to build things... Still investigating as I build a minimal app to reproduce things...

**Comment by tholo** [comment]
The minimal app required to reproduce the problem looks like:

```python
from pyramid.config import Configurator
from pyramid.view import view_config


@view_config(route_name='users', renderer='json')
def get_users(request):
    return []


def main(global_config, **settings):
    """This function returns a Pyramid WSGI application."""
    settings['pyramid_swagger.exclude_paths'] = [
        r'^/swagger.json',
    ]

    config = Configurator(settings=settings)
    config.include('pyramid_swagger')
    config.include(__name__, route_prefix='/api/v1')
    return config.make_wsgi_app()


def includeme(config):
    config.add_route('users', '/users')
    config.scan()
```

Requirements for that looks like:

```text
pyramid
pyramid_swagger
waitress
```

**Comment by tholo** [comment]
So, this appears to actually be a Pyramid issue. Further simplification of test:

```python
from pyramid.config import Configurator


def main(global_config, **settings):
    """This function returns a Pyramid WSGI application."""
    config = Configurator(settings=settings)
    config.add_route('home', '/some/path')
    config.include(__name__, route_prefix='/api/v1')
    return config.make_wsgi_app()


def includeme(config):
    config.add_route('users', '/users')


if __name__ == '__main__':
    app = main({}, settings={})

    print(app.routes_mapper.routes['home'].path)
    print(app.routes_mapper.routes['users'].path)
```

And the result of executing that is:

```text
/some/path
api/v1/users
```

**Comment by tholo** [comment]
This is now reported as Pylons/pyramid#2143

**Comment by striglia** [comment]
Awesome. Thanks for all the digging here, and hopefully that ticket will solve the issue.

If you can think of any generally useful improvements we can make on the pyramid_swagger side to clear up issues like this, feel free to file an issue.

**Comment by djeebus** [comment]
I believe this was fixed on our side in #151. Can anyone confirm?

**Comment by tholo** [comment]
That was almost exactly the same as my workaround, and does indeed fix the issue for me.
